### PR TITLE
*: adjust hot scheduler

### PR DESCRIPTION
**Pull request by nolouch**

### CODE REVIEW

File: server/core/store.go [hunk 1]
```diff
@@ -302,6 +302,15 @@ func (s *StoresInfo) GetStore(storeID uint64) *StoreInfo {
 	return store.Clone()
 }
 
+// ZoreGetStore return a StoreInfo with storeID
+func (s *StoresInfo) ZoreGetStore(storeID uint64) *StoreInfo {
```
Comment (siddontang) [inline server/core/store.go]: what does Zore mean?

File: server/schedule/hot_cache.go [hunk 2]
```diff
@@ -142,30 +179,40 @@ func (w *HotSpotCache) isNeedUpdateStatCache(region *core.RegionInfo, hotRegionT
...
+		cache, ok := w.readFlow[item.StoreID]
+		if !ok {
+			panic("no cache create!")
+		}
 		if item == nil {
-			w.readFlow.Remove(key)
+			cache.Remove(key)
 		} else {
-			w.readFlow.Put(key, item)
+			cache.Put(key, item)
```
Comment (disksing) [inline server/schedule/hot_cache.go]: I think you can extract a function `GetCache(flowKind, storeID) cache.Cache`.

File: server/schedule/hot_cache.go [hunk 3]
```diff
@@ -175,15 +222,28 @@ func (w *HotSpotCache) RegionStats(kind FlowKind) []*core.RegionStat {
...
+	len := len(stats)
+	if len > totalCacheLen {
+		len = totalCacheLen
 	}
-	return stats
+	res := make([]*core.RegionStat, len)
+	for i := 0; i < len; i++ {
+		res[i] = &stats[i]
+	}
+	return res
```
Comment (disksing) [inline server/schedule/hot_cache.go]: `return stats[:totalCacheLen]`

File: server/schedule/basic_cluster.go [hunk 4]
```diff
@@ -27,7 +27,8 @@ const (
 	// RegionHeartBeatReportInterval is the heartbeat report interval of a region
 	RegionHeartBeatReportInterval = 60
 
-	statCacheMaxLen              = 1000
+	statCacheMaxLen              = 100
```
Comment (disksing) [inline server/schedule/basic_cluster.go]: `cacheLenPerStore`.

File: server/core/store.go (hunk 1 above)
Comment (nolouch) [reply server/core/store.go]: :open_mouth: mean do not clone in here, has changed the name.

File: server/schedule/hot_cache.go [hunk 5]
```diff
@@ -175,15 +228,28 @@ func (w *HotSpotCache) RegionStats(kind FlowKind) []*core.RegionStat {
...
+		stats[i] = *elements[i].Value.(*core.RegionStat)
 	}
-	return stats
+	sort.Sort(core.RegionsStat(stats))
+	len := len(stats)
+	if len > totalCacheLen {
+		len = totalCacheLen
+	}
+	res := make([]*core.RegionStat, len)
+	for i := 0; i < len; i++ {
```
Comment (Connor1996) [inline server/schedule/hot_cache.go]: `i < math.Min(len(stats), totalCacheLen)` would be more clear

File: server/schedule/hot_cache.go [hunk 6]
```diff
@@ -198,13 +264,17 @@ func (w *HotSpotCache) RandHotRegionFromStore(storeID uint64, kind FlowKind, hot
 }
 
 func (w *HotSpotCache) isRegionHot(id uint64, hotThreshold int) bool {
-	if stat, ok := w.writeFlow.Peek(id); ok {
-		if stat.(*core.RegionStat).HotDegree >= hotThreshold {
+	stats := w.RegionStats(WriteFlow)
+	for _, stat := range stats {
+		if stat.RegionID == id && stat.HotDegree > hotThreshold {
```
Comment (Connor1996) [inline server/schedule/hot_cache.go]: not same condition as below, `>` and `>=`

File: server/schedule/hot_cache.go [hunk 7]
```diff
@@ -142,30 +172,53 @@ func (w *HotSpotCache) isNeedUpdateStatCache(region *core.RegionInfo, hotRegionT
...
+func (w *HotSpotCache) Update(origin, region *core.RegionInfo, item *core.RegionStat, kind FlowKind) {
+	needDelete := origin != nil && origin.Leader != nil && origin.Leader.GetStoreId() != region.Leader.GetStoreId()
 	switch kind {
 	case WriteFlow:
+		if needDelete {
+			if oldCacheFlow, ok := w.writeFlow[origin.Leader.GetStoreId()]; ok {
+				oldCacheFlow.Remove(origin.GetId())
+			}
+		}
+		cacheFlow, ok := w.writeFlow[item.StoreID]
```
Comment (Connor1996) [inline server/schedule/hot_cache.go]: item may be nil

**Pull request closed without merging**